### DISCUSSION TITLE
The future of `globalPreload`

### DISCUSSION THREAD

**Discussion by GeoffreyBooth** [nodejs/loaders#124] [Q&A] [answered]
In working on https://github.com/nodejs/node/pull/44710, moving loaders off-thread, I’ve run into various issues related to the `globalPreload` hook. It has an undocumented `setImportMetaCallback` function available alongside the `getBuiltin` and `port` globals that are provided for use in the bespoke sloppy script environment for use by `globalPreload` code; both `setImportMetaCallback` and `port` are problematic for moving loaders off thread, because `import.meta.resolve` needs to be defined in the main thread as far as I can tell, and is the `port` supposed to provide a link into the loaders thread or into the module loader that lives on the main thread? All of these issues _might_ be solvable, but I’m starting to wonder if they’re worth the effort.

Starting with just `setImportMetaCallback` and `port`, these were added in https://github.com/nodejs/node/pull/39240, which aimed to prove that mocking was possible with loaders. As far as I can tell, neither of these features have been used anywhere else or built on at all; and mocking already has solutions with broad adoption, like https://github.com/testdouble/testdouble.js/ and https://github.com/testdouble/quibble. Here’s the [`globalPreload` use in Quibble](https://github.com/testdouble/quibble/blob/5d927130f42b518498fa87cd36c0566755a9c79c/lib/quibble.mjs#L9-L15), one of the very few uses of `globalPreload` that I’ve been able to find in the wild:

```js
function loadQuibble () {
  global.__quibble = { stubModuleGeneration: 1 }
}

export function globalPreload () {
  return `(${loadQuibble.toString()})()`
}
```

It doesn’t use `setImportMetaCallback` and `port`, which leads me to believe that either the new features aren’t necessary for mocking or they don’t enable any improvements over existing solutions. Whereas conversely, they add a lot of complexity and a significant maintenance burden. I think that we should remove them, at least until after https://github.com/nodejs/node/pull/44710 lands; and any PR to re-add them needs to justify why they’re needed and better than existing solutions.

More broadly, I’m wondering why we need `globalPreload` at all. It’s a _weird_ API: [sloppy mode scripts, neither ESM nor CommonJS, that can’t import or require dependencies other than Node builtins](https://nodejs.org/api/esm.html#globalpreload). The only examples or usage that I can find set a few trivial global-scope variables like counters. Couldn’t the same be achieved by prepending some code to the `source` returned by the `load` hook?

Looking at [the original PR that created `globalPreload`, then called `getGlobalPreloadCode`](https://github.com/nodejs/node/pull/32068), the justifications I see given are:

> This work is meant to unblock progress on moving hook execution out of the main thread (and global scope). It creates an official place where hooks may interact with the global scope, hopefully allowing us to remove the need for random global side effects elsewhere. . . .
>
> One fundamental design decision implied here is: there are multiple people who want to move loader hooks to run on a separate thread.
>
> The use cases are instrumenting core modules and setting up any kind of global state required by code generated by other hooks. E.g. if transformSource assumes the existence on certain global runtime functions. Another use case is global attenuation / locking down the runtime before user code runs.
>
> Right now loader hooks may just rely on the fact that they happen to run in the same global scope as all other code but when moving to threads, that is no longer true.

The security concerns are already much better handled by a few other in-progress APIs under https://nodejs.org/api/permissions.html. The “define things in global scope” like polyfills or helpers can be done via `--import`, which is a much better way to do so because `--import` allows real ESM code that can actually import dependencies. I would think that loaders could also prepend whatever they wanted to the `source` returned by `load` for the application entry point (the only resolved module whose `parentURL` is undefined). Or put another way, I don’t know what `globalPreload` provides for us that can’t be achieved in better ways via other APIs.

Of course, I’m inviting everyone to now go ahead and prove me wrong, so feel free. 😄 I think if we do need to keep `globalPreload` for some reason, in the near term I think we have to at least revert https://github.com/nodejs/node/pull/39240 at least temporarily to get the off-thread PR finished and move the overall Loaders API to stable; but we would mark `globalPreload` specifically as still experimental, as we continue to investigate its use cases and what extra features it needs like communication channels.

cc @nodejs/loaders @JakobJingleheimer @jkrems @bmeck @giltayar

**Comment by giltayar** [comment, marked as answer]
I currently don't use `port` because the loader works "on thread", and so I can communicate via global variables.

But I already changed what needs to change to support "off thread" work, and you can see the `globalPreload` code which uses `port` here: https://github.com/testdouble/quibble/blob/support-off-thread-loaders/lib/quibble.mjs#L128

So, yes: `globalPreload` with `port` is necessary exactly _for_ when loaders become "off thread".

**Reply by GeoffreyBooth** [reply]
@giltayar Would you mind reducing https://github.com/testdouble/quibble/blob/support-off-thread-loaders/lib/quibble.mjs#L128 to a minimal unit test of the `port` feature? That has as little else as possible; just a unit test of `port`. I’d like a test in the Node codebase for `port` in isolation, basically the inverse of https://github.com/nodejs/node/blob/main/test/fixtures/es-module-loaders/mock-loader.mjs which is a test of a mocking reference implementation.

**Reply by giltayar** [reply]
@GeoffreyBooth Would love to (assuming it's still relevant), but it'll take me a week or two because I _gotta_ finish writing this talk that I'm supposed to present in two weeks.

**Comment by giltayar** [comment]
In other words, and if I was not clear: if you move loaders off thread, and don't give `quibble/testdouble` some kind of channel to communicate with the main thread, then it cannot be made to work "off thread".

**Reply by GeoffreyBooth** [reply]
Why does this need to be in `globalPreload` as opposed to being part of what's returned by `load`?

And why does Quibble need to communicate between threads at all?

**Reply by giltayar** [reply]
> And why does Quibble need to communicate between threads at all?

There are two parts to quibble: the API, which is used by the testing code to specify what module needs to be mocked and how (e.g. `quibble.esm('foo-library', {namedExport: mockingFunction})`. This API needs to communicate this "mocking request" by the user to the second part: the Quibble loader, which is in (or actually, will be in) another thread, and which does the mocking itself by using `resolve` and `load`.

> Why does this need to be in globalPreload as opposed to being part of what's returned by load?

`globalPreload` has nothing to do with `load`. It's an "init" function called by Node.js to send some stuff to the loader. The important stuff is the message channel port which enables the loader to communicate with the main thread.

**Reply by GeoffreyBooth** [reply]
So the idea here is for main thread application code to essentially provide configuration information (what to mock) to the Quibble loader, after the application has begun? So that future resolved modules can be mocked? So the important part is that the channel talks into the worker thread, I'd think.

What is the purpose of the code injected into each module?

**Reply by giltayar** [reply]
> So the idea here is for main thread application code to essentially provide configuration information (what to mock) to the Quibble loader, after the application has begun? So that future resolved modules can be mocked? So the important part is that the channel talks into the worker thread, I'd think.

Yes, main -> loader thread. But there's also the other way around (loader => main). When the main thread calls `quibble.esm`, it needs to wait till the worker thread finishes the request. I do this also via the bi-directional message channel. You can see it in the code I linked to above.

> What is the purpose of the code injected into each module?

The code is essentially an export for each export of the original module. And each export is a reference to the mocking provided by the user of quibble.

**Reply by giltayar** [reply]
This is not mocking-specific. Any loader that has an API that can affect it will need this.

**Reply by arcanis** [reply]
If I understand correctly it'd also allow to experiment in userland with dynamic loaders? That seems an interesting use case

**Reply by giltayar** [reply]
Yes. You could theoretically build a message channel protocol that can proxy the `resolve` and `load` hooks into userland.

But that would defeat the whole idea of moving to a worker thread. And I'm guessing the performance wouldn't be very good.

**Reply by GeoffreyBooth** [reply]
> Yes, main -> loader thread. But there's also the other way around (loader => main). When the main thread calls `quibble.esm`, it needs to wait till the worker thread finishes the request. I do this also via the bi-directional message channel. You can see it in the code I linked to above.

So in the off-thread PR world, there are two “internal” places the communications port can connect to: inside the worker thread where `resolve` and `load` are run, or inside Node internals where ESMLoader lives, on the main thread. I assume you want to communicate into the worker thread?

One of the things that seems to need to be done on the main thread is defining `import.meta`. Maybe that’s a limitation that we can overcome in the future, but I’ve spent a lot of time on it and I can’t get it working within the worker thread. I think the only consequence of this is that I’ll need to remove `setImportMetaCallback`. The `import.meta.resolve` function needs to be called from the main thread because that’s how we get the sync `resolve` that calls into the thread’s async `resolve` chain.

I’ll try to preserve `port`.

> 
> > What is the purpose of the code injected into each module?
> 
> The code is essentially an export for each export of the original module. And each export is a reference to the mocking provided by the user of quibble.

What difference does it make to inject this via `globalPreload` instead of prepending the source returned by `load`? Assuming the `port` was provided to the `load` hook like it currently is to `globalPreload`. I can imagine it might affect source maps, and it would make debugging a bit messier as the loader-injected code would appear before the user’s code, but other than cosmetic things?

**Reply by giltayar** [reply]
> I assume you want to communicate into the worker thread?

As I said, I want to communicate into the worker thread, and also from the worker thread to the other side (it is bidirectional) so that I can tell the other side that the API call is done.

> and I can’t get [`import.meta`] working within the worker thread

That's a limitation I can live with for the moment. But obviously, for the sake of symmetry, it needs to be resolved (no pun intended 😂).

> The `import.meta.resolve` function needs to be called from the main thread because that’s how we get the sync resolve that calls into the thread’s async resolve chain.

Yes. Definitely. `import.meta.resolve` should not be called by any loader. Otherwise... I can't even think of the weird recursion that can happen. But this is a limitation loaders can live with, because (i) it's absolutely not a problem to relegate all calls to `import.meta.resolve` to the main thread, and (ii) you're planning on giving helper functions that one can use instead of `import.meta.resolve`?

> What difference does it make to inject this via `globalPreload` instead of prepending the source returned by `load`? Assuming the port was provided to the load hook like it currently is to `globalPreload`

Injecting it in the first call to `load` would work, but it's really really weird. First off, I would have to check that this is the first call to load and try and do it only once (not to mention dealing with the race conditions of two loads happening at the same time). And, secondly, the user _could_ call the `quibble.esm(...)` API before even _one_ module has loaded, and so the injected code that is used by `quibble.esm(...)` won't be there to be used by it.

It doesn't make sense.

**Reply by hybrist** [reply]
One of the design constraints was that the code should be [allowed to be] in sloppy mode. I forgot the exact case where it mattered though - maybe access to the always and reliably global this? Prepending to existing code, especially module code, may make this hard or even impossible.

**Reply by ljharb** [reply]
Polyfills definitely need that capability, which loaders need to be able to then inject. Not sure if that means there's a constraint here or not tho.

**Reply by GeoffreyBooth** [reply]
> One of the design constraints was that the code should be [allowed to be] in sloppy mode. I forgot the exact case where it mattered though - maybe access to the always and reliably global this? Prepending to existing code, especially module code, may make this hard or even impossible.

I thought the sloppy mode was just because that’s all we could create with `vm`? Wouldn’t ESM code always have access to `globalThis`? Yes I know it could be deleted or something, but I think if user code did that and then a loader didn’t work, it shouldn’t be surprising. I don’t consider loaders to be a security feature.

> Injecting it in the first call to `load` would work, but it’s really really weird. First off, I would have to check that this is the first call to load and try and do it only once (not to mention dealing with the race conditions of two loads happening at the same time). And, secondly, the user _could_ call the `quibble.esm(...)` API before even _one_ module has loaded, and so the injected code that is used by `quibble.esm(...)` won’t be there to be used by it.

I think realistically it would be injected via `--import`, as that would be processed before any user code and wouldn’t make debugging application code messy. It would mean the users would need to be instructed to pass two flags instead of one, which is unfortunate but I think the solution there is a config file for Node options, which is something else on our to-do list. The advantage of `--import` (or prepending to `load` source) is that we’re in a normal ESM environment where you could `import` dependencies, not just Node builtins.

Or put another way, what does `globalPreload` provide for us that `--import` doesn’t? The `port` communications channel, certainly, but couldn’t that be provided in some simpler fashion like `import { loadersThreadPort } from 'module'` or something? (And in fairness to @jkrems, `globalPreload` was created years before `--import`.)

Is the other difference is that the script returned by `globalPreload` is run before (or as part of) _every_ module, not just before the first one? So code in `--import` runs once, but `globalPreload` runs as many times as there are modules loaded for an application?

**Reply by hybrist** [reply]
> I thought the sloppy mode was just because that’s all we could create with vm?

No, we could've created a strict mode environment if we wanted to. But the explicit goal was to *not* force the code into strict mode since it's ~impossible to escape if sloppy mode is needed.

The two independent runtime environment things in the global preload were:
1. Access to a communication channel with the loader thread.
2. Access to sloppy mode.

Without leaking the port to the application code, I'm not sure if (1) can be done via code injection. (2) may need a more concrete use case and even then it may be arguable that other mechanisms like vm could be used to inject sloppy code even in a strict global preload environment.

**Reply by giltayar** [reply]
> Is the other difference is that the script returned by globalPreload is run before (or as part of) every module?

No. Once only.

> Or put another way, what does globalPreload provide for us that --import doesn’t?

Access to a bidirectional port to communicate with the worker thread. 

There's one more thing: `globalPreload` needs to run _for every worker thread spawned by the user_. Because the loader thread needs to communicate with all of them.

(And, yes, Quibble's code doesn't yet handle that. But it should.)

**Reply by GeoffreyBooth** [reply]
> Access to a bidirectional port to communicate with the worker thread.
> 
> There’s one more thing: `globalPreload` needs to run _for every worker thread spawned by the user_. Because the loader thread needs to communicate with all of them.

Let’s step back for a second, though. When `globalPreload` was introduced in https://github.com/nodejs/node/pull/32068 in March 2020, it didn’t have a communications channel. The `port` helper was added in https://github.com/nodejs/node/pull/39240 in July 2021. So is it fair to say that the _original_ `globalPreload` didn’t do anything that `--import` can do instead? (Leaving aside the sloppy mode consideration, since we can’t seem to pin down why that was desired; I also feel like it should be doable within `--import`, either through `vm` like @jkrems suggests or maybe by ESM code importing a CommonJS file.)

Because if `--import` can do everything that the injected sloppy mode script can do, it begs the question whether an injected script is the best way to provide a hook to a communications channel. Maybe we could leave `--import` for the “inject code before user application code” responsibility and create a new API specifically to provide a communications channel between the loaders thread and the main thread. What would be the downside of doing so?

As for any of this running for each worker thread spawned by the user, that’s a whole other ball of complexity that I’m not sure we’ve ever tested. Arguably we might want a way to run `--import` for each worker thread, for example to add instrumentation or polyfills to each worker thread, beyond just the use case of loader thread communication.

**Reply by giltayar** [reply]
`--import` would work (assuming it should run per thread, and I think it should), but:

1) I think it is not good DX to now require _two_ options to make this kind of thing work, and.
2) You still need to find some way to communicate via a port.

**Reply by hybrist** [reply]
For import: I assume the port would exist per loader since multiple loaders shouldn't have to worry about stepping on each others messages. How would this be done if the code is injected via a separate flag?

**Reply by arcanis** [reply]
There's some precedent for having a single message channel that all consumers share and filter as needed (that's how iframes and workers behave, through postmessage)

**Reply by GeoffreyBooth** [reply]
> `--import` would work (assuming it should run per thread, and I think it should), but:
> 
> 1. I think it is not good DX to now require _two_ options to make this kind of thing work, and.
> 2. You still need to find some way to communicate via a port.

Just so we don’t skip over the question, though, _does_ `globalPreload` provide anything that `--import` or `--require` don’t other than the communication channel? Regardless of whether we think whatever non-`globalPreload` option is better DX, what functionality would be lost if we removed the `globalPreload` hook entirely and instead provided some other way of establishing a communication channel? I’m not saying that `--import` plus who-knows-what _should_ be the way that that’s done; it’s just a hypothetical to tease out what value we’re getting out of `globalPreload`. Once I know what functionality we need to preserve, then we can discuss what API is the best way to achieve it, both from a DX perspective and a maintainability perspective (which might very well be `globalPreload`, but we need to sketch out all our options to be sure).

**Reply by giltayar** [reply]
@GeoffreyBooth all I need is a bidirectional communication channel to each user thread (or, as @arcanis said), a shared one for all user threads. I'm guessing I could differentiate based on thread id.

**Reply by GeoffreyBooth** [reply]
> @GeoffreyBooth all I need is a bidirectional communication channel to each user thread (or, as @arcanis said), a shared one for all user threads. I’m guessing I could differentiate based on thread id.

That’s fine. What I’m getting at is that I don’t know if the current design is the most sensible one. Like even if we want sloppy mode scripts injected into each thread, maybe that’s not _also_ the best place to define the communications channel. They don’t seem necessarily connected.

Another API we’ve been discussing is `import { registerLoader } from 'module'`, to allow registering loaders from application code. Such loaders would only apply to modules imported after the loaders are registered, so it would be used something like `await registerLoader('my-loader'); await import('application-entry-point.js');` but this would allow you to put any other application-side initialization code you wanted there, such as to define globals or register communications channels and so on. Again I’m not sure if this is better than any of the alternatives, but just an option for discussion.

**Reply by giltayar** [reply]
> That’s fine. What I’m getting at is that I don’t know if the current design is the most sensible one. 

I agree. Having to send over some code that is eval-ed in userland doesn't seem like it's the best API. One could imagine an additional function in the loader (`getUserLandModule` and please ignore the really bad name) that accepts a port and returns a module specifier for that loader that is loaded in userland (as opposed to the loader thread), and which contains an exported function that it contains a port. E.g.

```js
// loader.mjs

function getUserLandModule(portThatCommunicatesWithTheUserLandModuleReturned) {
  // save the port somewhere

  return './user-land-part-of-loader.mjs';
}

// user-land-part-of-loader.mjs

function init(portThatCommunicatesWithLoader) {
 // save the port somewhere
}
```

**Comment by bmeck** [comment]
To an extent yes userland loaders could use this, but things like multi-thread coordination and head of line blocking would be pretty rough.

There is a test for this exact kind of mocking in core already at https://github.com/nodejs/node/blob/main/test/fixtures/es-module-loaders/mock-loader.mjs that relies on the message channel. It was undocumented when implemented as it isn’t something that gives any benefits unless loaders are moved to their own thread.

**Reply by GeoffreyBooth** [reply]
Yes, this is the test I’m struggling to get to pass. I think I need to refactor it to remove `setImportMetaCallback` and the parts that depend on `import.meta`, which are pretty central to the test so I might have to remove it completely and it gets re-added in a follow-up PR.

This appears to be the only test in the codebase that tests `globalPreload` `port`, but it simultaneously tests a _lot_ of other things. What would really help is a much simpler test that tests `port` by itself in isolation. cc @giltayar

**Comment by bmeck** [comment]
Sloppy mode was as @ljharb stated. Polyfills. It is a hook to modify the
runtime not for module loading

**Comment by bmeck** [comment]
ServiceWorker is closer to loaders than that. They do not do broadcast. 1
to 1 communication is the norm there. Multi response is an issue

**Comment by bmeck** [comment]
Sloppy mode and guarantee of execution on first tick.

**Reply by GeoffreyBooth** [reply]
I think `--require` provides both of those?

**Comment by bmeck** [comment]
Yep but requires disk unlike import due to using paths

**Comment by Flarna** [comment]
We could add a possibility to register loaders programmatically in a `--import` or `--require` script to avoid the need that users have to provide two command line arguments to have the globalPreload part and the loader part in place.

**Reply by GeoffreyBooth** [reply]
Yes, this is what I was referring to when I mentioned `import { registerLoader } from 'module'` somewhere. The potential issue with such an API is dependencies that might call it, which is an issue we already have with `require.extensions` and so maybe it isn’t a concern.

**Comment by GeoffreyBooth** [comment]
Once we finish shaking out the bugs from moving loaders off the main thread, I think the `resolve` and `load` hooks will probably be ready to be marked stable (with the exception of `context.importAssertions` within each hook). For `globalPreload`, though, I would like to do a redesign probably along the lines of https://github.com/nodejs/loaders/discussions/124#discussioncomment-4554457, where the “sloppy script” aspect goes away and all that the API does is provide a communication port between application code and a particular loader, similar to https://nodejs.org/api/worker_threads.html#class-messagechannel. What do people think of this?
